riscv/csrs.h

```diff
@@ -500,4 +500,68 @@ class hgatp_csr_t: public basic_csr_t {
 };
 
 
+class tselect_csr_t: public basic_csr_t {
+ public:
+  tselect_csr_t(processor_t* const proc, const reg_t addr);
+ protected:
+  virtual bool unlogged_write(const reg_t val) noexcept override;
+};
+
+
+class tdata1_csr_t: public csr_t {
+ public:
+  tdata1_csr_t(processor_t* const proc, const reg_t addr);
+  virtual reg_t read() const noexcept override;
+ protected:
+  virtual bool unlogged_write(const reg_t val) noexcept override;
+};
+
+class tdata2_csr_t: public csr_t {
+ public:
+  tdata2_csr_t(processor_t* const proc, const reg_t addr, const size_t count);
+  virtual reg_t read() const noexcept override;
+  reg_t read(const size_t idx) const noexcept;
+ protected:
+  virtual bool unlogged_write(const reg_t val) noexcept override;
+ private:
+  std::vector<reg_t> vals;
+};
+
+// For CSRs that are only writable from debug mode
+class debug_mode_csr_t: public basic_csr_t {
+ public:
+  debug_mode_csr_t(processor_t* const proc, const reg_t addr);
+  virtual void verify_permissions(insn_t insn, bool write) const override;
+};
+
+typedef std::shared_ptr<tdata2_csr_t> tdata2_csr_t_p;
+
+
+class dpc_csr_t: public epc_csr_t {
```

Comment:
I assume this does not inherit debug_mode_csr_t to avoid multiple inheritance.

Reply:
Yes. The disappointing thing is that without using multiple inheritance (and diamond-pattern multiple inheritance) I have to have three separate implementations of `verify_permissions()` that are all identical (for `debug_mode_csr_t`, `dpc_csr_t`, and `dcsr_csr_t`).

Do you think there's any good way to avoid that? I haven't researched C++ multiple inheritance in decades.

Reply:
I do not. Personally I don't understand the details of multiple inheritance well enough so I stay away from it.